source/common/router/router.h

```diff
@@ -198,7 +198,12 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     return callbacks_->connection();
   }
   const Http::HeaderMap* downstreamHeaders() const override { return downstream_headers_; }
-
+  absl::optional<std::function<bool(uint32_t, const Upstream::Host&)>>
```

Comment:
Also, for this function in particular, what are you thoughts on performance? Obviously this is not going to be free for people that use this. What scenarios do you imagine the pre-filter being used in? I'm trying to get a picture of what types of filters people will build.

Reply:
When I was toying around with this internally, the filters would generally just compare some attribute of the host with previously attempted hosts. Given a vector `attempted_hosts`, the predicates generally looked like:

explicitly target different zone after two attempts
```
if (attempted_hosts.size() > 2) {
  [&attempted_hosts](auto, const Host& host) {
    for (const auto& attempted : attempted_hosts) {
      if (host.locality()->zone() == attempted.locality()->zone();
    }
  }
}
```

always target different zone from previous attempt
```
[&attempted_hosts](auto, const Host& host) {
    return host.locality()->zone() == std::prev(attempted_hosts.end())->locality()->zone(); 
}
```

Essentially I'm imagining the returned predicate to compare some string attribute of the host, be that locality or some arbitrary metadata, with one or more previously attempted hosts. Possibly we could get away with pointer equality for filtering out specific hosts from the list of hosts that were previously attempted.